adapter-manager.js

```diff
@@ -13,38 +14,245 @@
 var fs = require('fs');
 var path = require('path');
 var EventEmitter = require('events').EventEmitter;
+var Deferred = require('./deferred');
 
+/**
+ * @class AdapterManager
+ * @classdesc The AdapterManager will load any adapters from the 'adapters'
+ * directory. See loadAdapters() for details.
+ */
 class AdapterManager extends EventEmitter {
 
   constructor() {
     super();
     this.adapters = [];
     this.devices = {};
+    this.deferredAdd = null;
+    this.deferredRemove = null;
   }
 
+  /**
+   * Adds an adapter to the collection of adapters managed by AdapterManager.
+   * This function is typically called when loading adapters.
+   */
   addAdapter(adapter) {
     adapter.name = adapter.constructor.name;
     this.adapters.push(adapter);
+
+    /**
+     * Adapter added event.
+     *
+     * This is event is emitted whenever a new adapter is loaded.
+     *
+     * @event adapter-added
+     * @type  {Adapter}
+     */
     this.emit('adapter-added', adapter);
   }
 
-  addDevice(adapter, device) {
+  /**
+   * @method deviceAdded
+   *
+   * Called when the indicated device has been added to an adapter.
+   */
+  deviceAdded(device) {
     this.devices[device.id] = device;
+
+    /**
+     * Device added event.
+     *
+     * This event is emitted whenever a new device is added.
+     *
+     * @event device-added
+     * @type  {Device}
+     */
     this.emit('device-added', device);
+
+    var deferredAdd = this.deferredAdd;
+    if (deferredAdd) {
+      this.deferredAdd = null;
+      for (var adapter of this.adapters) {
+        if (adapter !== device.adapter) {
+          adapter.cancelPairing();
+        }
+      }
+      deferredAdd.resolve(device);
+    }
+  }
+
+  /**
+   * @method addSomeDevice
+   * Initiates pairing on all of the adapters that support it.
+   * The user then presses the "button" on the device to be added.
+   * @returns A promise that resolves to the newly added device.
+   */
+  addSomeDevice() {
```

Comment:
Yeah - I renamed addDevice to deviceAdded. Maybe I should call it _deviceAdded. It's not really part of the public facing API of the adapter, but rather a class private function which the particular adapters should call.

I didn't want to call it onDeviceAdded because it wasn't hooked up as a listener.